fuel-merkle/src/binary/in_memory.rs

```diff
@@ -62,6 +67,72 @@ impl Default for MerkleTree {
     }
 }
 
+#[derive(Default, Debug, Clone)]
+pub struct MerkleRootCalculator {
+    stack: Vec<Node>,
+}
+
+impl MerkleRootCalculator {
+    pub fn new() -> Self {
+        Self { stack: Vec::new() }
+    }
+
+    pub fn push(&mut self, data: &[u8]) {
+        let node = Node::create_leaf(0, data);
+        self.stack.push(node);
+
+        if let Some(top_node) = self.stack.last() {
+            if self.stack.len() > 1 {
+                if let Some(second_top_node) = self.stack.get(self.stack.len() - 2) {
+                    if second_top_node.height() == top_node.height() {
+                        let merged_node = Node::create_node(second_top_node, top_node);
+                        self.stack.pop();
+                        self.stack.pop();
+                        self.stack.push(merged_node);
+                        while let Some(last_merged) = self.stack.last() {
+                            if self.stack.len() > 1 {
+                                if let Some(second_last_merged) =
+                                    self.stack.get(self.stack.len() - 2)
+                                {
+                                    if last_merged.height() == second_last_merged.height()
+                                    {
+                                        let merged_node = Node::create_node(
+                                            second_last_merged,
+                                            last_merged,
+                                        );
+                                        self.stack.pop();
+                                        self.stack.pop();
+                                        self.stack.push(merged_node);
+                                    } else {
```

Comment:
You duplicate the same logic before this loop. I think you can have one loop that does these checks.

Reply:
You're absolutely right! I overlooked this. 😅 At one point I needed both...